### DISCUSSION TITLE
Automatic light/dark colorscheme

### DISCUSSION THREAD

**Discussion by csoto-3** [LazyVim/LazyVim#5201] [Extra Requests]
Would be nice to have my LazyVim automatically switch colorscheme based on the terminal/OS theme. Ideally it'd be configurable such that users can pick separate themes for light and dark mode (rather than toggle a single theme). I came across [auto-dark-mode.nvim](https://github.com/f-person/auto-dark-mode.nvim) but I can't figure out how to interface with LazyVim's colorscheming interface. Adding this plugin (or one like it) as a preconfigured extra seems like a good way to go about it.

**Comment by csoto-3** [comment]
I wrote a simple config for auto-dark-mode.nvim that seems to work. Hopefully this makes it easier to add as an extra:

```lua
return {
  "f-person/auto-dark-mode.nvim",
  lazy = false,
  priority = 1000,
  opts = {
    update_interval = 1000,
    set_dark_mode = function()
      vim.api.nvim_set_option_value("background", "dark", {})
      vim.cmd("colorscheme gruvbox-material")
    end,
    set_light_mode = function()
      vim.api.nvim_set_option_value("background", "dark", {})
      vim.cmd("colorscheme everforest")
    end,
  },
}
```

**Reply by SimonSimCity** [reply]
I've changed it to only update the `background` variable, as this is [what's needed for tokyonight](https://github.com/folke/tokyonight.nvim?tab=readme-ov-file#%EF%B8%8F-configuration) 🎉 

For those unfamiliar with neovim, you just need to create a new file e.g. called `auto-dark-mode.lua` in the folder `~/.config/nvim/lua/plugins/` with the following content:

```lua
return {
  "f-person/auto-dark-mode.nvim",
  lazy = false,
  priority = 1000,
  opts = {
    update_interval = 1000,
    set_dark_mode = function()
      vim.api.nvim_set_option_value("background", "dark", {})
    end,
    set_light_mode = function()
      vim.api.nvim_set_option_value("background", "light", {})
    end,
  },
}
```

Just this question: I see almost every suggestion herer call `vim.cmd("colorscheme <some-scheme>")`. Is it really necessary to call this if you're not changing the color scheme? For me it seems to work without - but are there downsides to it if the theme relies on the option, as `tokyonight` does?

**Comment by dpetka2001** [comment]
This plugin sets the following options depending on the OS theme
```lua
set_dark_mode = function()
      vim.api.nvim_set_option_value("background", "dark", {})
      vim.cmd("colorscheme gruvbox")
    end,
    set_light_mode = function()
      vim.api.nvim_set_option_value("background", "light", {})
      vim.cmd("colorscheme gruvbox")
    end,
```
There's already a mapping `<leader>ub` to toggle between dark/light theme that does the same thing except it's not automatic. Just use that? I believe such a plugin is out of scope, since the functionality already exists with the keybinding.

**Reply by AdrienLemaire** [reply]
What I would like is for neovim to auto-exec <leader>ub when the dark/light mode toggles. Running a job every second doesn't seem efficient.

**Reply by dpetka2001** [reply]
> Running a job every second doesn't seem efficient

Neither is having a plugin that does what's already built in, just for the sake of automating. You can of course integrate it in your own personal configuration. OP already gave an example configuration.

**Reply by tkolleh** [reply]
How to use `<leader>ub` to change background and theme. E.g. toggle `setLight()` or `setDark()`:
```lua
--
-- Utility functions
--
local function setDark()
  vim.api.nvim_set_option_value("background", "dark", {})
  vim.cmd("colorscheme github_dark_tritanopia")
end

local function setLight()
  vim.api.nvim_set_option_value("background", "light", {})
  vim.cmd("colorscheme github_light_default")
end
```

**Comment by tkolleh** [comment]
Sharing my config. I'd like to attach these functions to `<leader>ub`;

```lua
--
-- Utility functions
--
local function setDark()
  vim.api.nvim_set_option_value("background", "dark", {})
  vim.cmd("colorscheme tokyonight-moon")
end

local function setLight()
  vim.api.nvim_set_option_value("background", "light", {})
  vim.cmd("colorscheme github_light_high_contrast")
end

local function is_background_dark()
  local bg = vim.opt.background:get()
  local colorscheme = vim.g.colors_name

  -- Check if it's explicitly set to 'dark'.
  if bg == "dark" then
    return true
  end

  -- Check if it's explicitly set to 'light'.
  if bg == "light" then
    return false
  end

  -- If background is not set explicitly, check the color scheme
  if colorscheme then
    -- List of colorschemes known to be light
    local light_colorschemes = {
      "github_light_default",
      "github_light_high_contrast",
      "github_light_tritanopia",
    }

    for _, scheme in ipairs(light_colorschemes) do
      if colorscheme == scheme then
        return false
      end
    end
  end

  return false
end

  -- Configure LazyVim to load dark or light mode scheme
  {
    "LazyVim/LazyVim",
    opts = {
      colorscheme = function()
        if is_background_dark() then
          setDark()
        else
          setLight()
        end
      end,
    },
  },
```

**Reply by yusufaktepe** [reply]
With `Snacks.toggle`

```lua
Snacks.toggle({
    name = "Dark Colorscheme",
    get = function()
      return vim.api.nvim_get_option_value("background", {}) == "dark"
    end,
    set = function(state)
      if state then
        vim.api.nvim_set_option_value("background", "dark", {})
        vim.cmd("colorscheme tokyonight-moon")
      else
        vim.api.nvim_set_option_value("background", "light", {})
        vim.cmd("colorscheme github_light_high_contrast")
      end
    end,
  }):map("<leader>ub")

```

**Reply by csoto-3** [reply]
How can I implement this code snippet in a standalone lua file? (e.g. `plugins/colorscheme.lua`).
